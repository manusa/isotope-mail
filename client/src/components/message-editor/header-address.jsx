import React, {Component} from 'react';
import {translate} from 'react-i18next';
import PropTypes from 'prop-types';
import Autosuggest from 'react-autosuggest';
import {validateEmail} from '../../services/validation';
import mainCss from '../../styles/main.scss';

export class HeaderAddress extends Component {
  constructor(props) {
    super(props);
    this.inputRef = React.createRef();
    this.handleOnSuggestionChange = this.onSuggestionChange.bind(this);
    this.handleOnSuggestionsFetchRequested = this.onSuggestionsFetchRequested.bind(this);
    this.handleOnSuggestionsClearRequested = this.onSuggestionsClearRequested.bind(this);
    this.handleOnSuggestionSelected = this.onSuggestionSelected.bind(this);
    this.handleOnHeaderKeyDown = this.onHeaderKeyDown.bind(this);
    this.handleOnHeaderBlur = this.onHeaderBlur.bind(this);

    this.state = {
      value: '',
      suggestions: []
    };
  }

  render() {
    const {
      id,
      className, chipClassName, autoSuggestClassName, autoSuggestMenuClassName,
      label, addresses, onAddressRemove
    } = this.props;
    const {suggestions, value} = this.state;
    return (
      <div className={`${className} ${mainCss['mdc-menu-surface--anchor']}`} onClick={() => this.fieldClick()}
        onDragOver={e => e.preventDefault()} onDrop={e => this.onDrop(e, id)}>
        <label>{label}:</label>
        {addresses.map((address, index) => (
          <div key={index} className={`${chipClassName} ${mainCss['mdc-chip']}`}
            draggable={true}
            onDragStart={event => HeaderAddress.onAddressDragStart(event, id, address)}>
            <div className={mainCss['mdc-chip__text']}>{address}</div>
            <i onClick={() => onAddressRemove(id, address)} className={`material-icons ${mainCss['mdc-chip__icon']}
               ${mainCss['mdc-chip__icon--trailing']}`}>cancel</i>
          </div>
        ))}
        <Autosuggest
          suggestions={suggestions}
          ref={this.inputRef}
          inputProps={{
            id: id,
            // type: 'email', <- Chrome in combination with autosuggest has bug with backspace, must perform manual validation
            type: 'text',
            value: value,
            onChange: this.handleOnSuggestionChange,
            onKeyDown: this.handleOnHeaderKeyDown,
            onBlur: this.handleOnHeaderBlur
          }}
          getSuggestionValue={HeaderAddress.getSuggestionValue}
          renderSuggestion={HeaderAddress.renderSuggestion}
          onSuggestionsFetchRequested={this.handleOnSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.handleOnSuggestionsClearRequested}
          onSuggestionSelected={this.handleOnSuggestionSelected}
          theme={{
            container: `${autoSuggestClassName} `,
            suggestionsContainer: `${mainCss['mdc-menu']} ${mainCss['mdc-menu-surface']} ${autoSuggestMenuClassName}`,
            suggestionsContainerOpen: `${mainCss['mdc-menu-surface--open']}`,
            suggestionsList: `${mainCss['mdc-list']} ${mainCss['mdc-list--dense']}`,
            suggestion: mainCss['mdc-list-item'],
            suggestionHighlighted: mainCss['mdc-list-item--activated']
          }}
        />
      </div>
    );
  }

  fieldClick() {
    this.inputRef.current.input.focus();
  }

  /**
   * Clears any HTML 5 validation errors from the provided target.
   *
   * @param target
   */
  static clearValidation(target) {
    target.setCustomValidity('');
  }

  onSuggestionChange(event, {newValue}) {
    this.setState({value: newValue});
  }

  onSuggestionsFetchRequested({value}) {
    this.setState({suggestions: this.props.getAddresses(value)});
  }

  onSuggestionsClearRequested() {
    this.setState({suggestions: []});
  }

  onSuggestionSelected(event, {suggestionValue}) {
    this.setState({value: ''});
    this.props.onAddressAdd(this.props.id, suggestionValue);
    setTimeout(() => HeaderAddress.clearValidation(this.inputRef.current.input));
  }

  /**
   * Computes the value for the provided suggestion object, as suggestions are an array of strings, no computation is
   * required.
   *
   * @param suggestion object
   * @returns {string} value to render
   */
  static getSuggestionValue(suggestion) {
    return suggestion;
  }

  /**
   * Returns a component to be rendered in the suggestionList container of the Autosuggest component
   *
   * @param suggestionValue
   * @returns {*}
   */
  static renderSuggestion(suggestionValue) {
    return suggestionValue;
  }

  validateEmail(event) {
    const target = event.target;
    const error = validateEmail(target.value);
    if (error) {
      event.preventDefault();
      target.setCustomValidity(error);
      setTimeout(() => target.reportValidity());
      return false;
    }
    return true;
  }

  onHeaderKeyDown(event) {
    const target = event.target;
    HeaderAddress.clearValidation(target);
    if (event.key === 'Enter' || event.key === ';') {
      if (this.validateEmail(event)) {
        const id = target.id;
        const value = target.value.replace(/;/g, '');
        this.props.onAddressAdd(id, value);
        this.setState({value: ''});
        target.focus();
        event.preventDefault();
      }
    }
  }

  onHeaderBlur(event) {
    const target = event.target;
    if (target.value.length > 0) {
      if (this.validateEmail(event)) {
        this.props.onAddressAdd(target.id, target.value);
        this.setState({value: ''});
      }
    }
  }

  static onAddressDragStart(event, id, address) {
    event.stopPropagation();
    const payload = {id, address};
    event.dataTransfer.setData('application/json', JSON.stringify(payload));
  }

  onDrop(event, id) {
    event.preventDefault();
    const types = event.dataTransfer.types;
    if (types && Array.from(types).indexOf('application/json') >= 0) {
      const payload = JSON.parse(event.dataTransfer.getData('application/json'));
      if (id && id !== payload.id) {
        const fromId = payload.id;
        const address = payload.address;
        this.props.onAddressMove(fromId, id, address);
      }
    }
  }
}

HeaderAddress.propTypes = {
  t: PropTypes.func,
  id: PropTypes.string.isRequired,
  className: PropTypes.string,
  chipClassName: PropTypes.string,
  autoSuggestClassName: PropTypes.string,
  autoSuggestMenuClassName: PropTypes.string,
  addresses: PropTypes.arrayOf(PropTypes.string),
  label: PropTypes.string,
  getAddresses: PropTypes.func,
  onAddressAdd: PropTypes.func,
  onAddressRemove: PropTypes.func,
  onAddressMove: PropTypes.func
};

HeaderAddress.defaultProps = {
  className: '',
  chipClassName: '',
  autoSuggestClassName: '',
  autoSuggestMenuClassName: '',
  addresses: [],
  label: '',
  onAddressAdd: () => {},
  onAddressRemove: () => {},
  onAddressMove: () => {}
};

export default (translate()(HeaderAddress));
