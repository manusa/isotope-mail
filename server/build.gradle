buildscript {
	ext {
		springBootVersion = '2.0.4.RELEASE'
	}
	repositories {
		mavenCentral()
	}
	dependencies {
		classpath("org.springframework.boot:spring-boot-gradle-plugin:${springBootVersion}")
	}
}

apply plugin: 'java'
apply plugin: 'eclipse'
apply plugin: 'org.springframework.boot'
apply plugin: 'io.spring.dependency-management'

group = 'com.marcnuri.isotope'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = 1.8

repositories {
	mavenCentral()
}


dependencies {
	compile('org.springframework.boot:spring-boot-starter')
	compile('org.springframework.boot:spring-boot-starter-web')
	compile('org.springframework.boot:spring-boot-starter-hateoas')
	compile('com.sun.mail:javax.mail:1.6.1')
	testCompile('org.springframework.boot:spring-boot-starter-test')
	testCompile('com.marcnuri:spring-common-test:0.0.2')
}
