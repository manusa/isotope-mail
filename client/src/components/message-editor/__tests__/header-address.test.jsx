import React from 'react';
import {mount, shallow} from 'enzyme/build/index';
import {HeaderAddress} from '../header-address';

describe('HeaderAddress component test suite', () => {
  test('Snapshot render, should render HeaderAddress', () => {
    const props = {
      t: translation => (translation),
      id: 'I\'mDown',
      className: 'Business',
      chipClassName: 'FirstClassChips',
      addresses: ['name@email.com', 'other@e-mail.mail'],
      label: 'BlackLabel12yrs',
      onAddressAdd: () => {},
      onAddressRemove: () => {},
      onAddressMove: () => {}
    };
    const headerAddress = shallow(<HeaderAddress {...props} />);
    expect(headerAddress).toMatchSnapshot();
  });
  describe('Event tests', () => {
    test('fieldClick, input grabs focus', () => {
      // Given
      const props = {id: 'to'};
      const focus = jest.fn();
      const headerAddress = mount(<HeaderAddress {...props}/>);
      headerAddress.find('input#to').getDOMNode().focus = focus;

      // When
      headerAddress.find('div').first().simulate('click', {});

      // Then
      expect(focus).toHaveBeenCalledTimes(1);
    });
    describe('Autosuggest built-in integrated events', () => {
      test('onSuggestionChange, non-empty value, state value is changed, and addresses are fetched', () => {
        // Given
        const getAddresses = jest.fn(() => ['test@test.com']);
        const props = {id: 'to', getAddresses};
        const headerAddress = mount(<HeaderAddress {...props}/>);

        // When
        headerAddress.find('input#to').simulate('change', {target: {value: '1337'}});

        // Then
        expect(getAddresses).toHaveBeenCalledTimes(1);
        expect(headerAddress.state().value).toEqual('1337');
        expect(headerAddress.state().suggestions).toEqual(['test@test.com']);
      });
      test('onSuggestionChange, empty value, addresses are cleared', () => {
        // Given
        const getAddresses = jest.fn(() => ['test@test.com']);
        const props = {id: 'to', getAddresses};
        const headerAddress = mount(<HeaderAddress {...props}/>);
        headerAddress.state().suggestions = ['not@empty'];

        // When
        headerAddress.find('input#to').simulate('change', {target: {value: ''}});

        // Then
        expect(headerAddress.state().suggestions).toEqual([]);
      });
      test('getSuggestionValue, valid string input, should return same input', () => {
        // Given
        const props = {id: 'to'};
        const headerAddress = shallow(<HeaderAddress {...props}/>);

        // When
        const result = headerAddress.find('Autosuggest').props().getSuggestionValue('1337');

        // Then
        expect(result).toBe('1337');
      });
      test('renderSuggestion, valid string suggestionValue, should return just a string', () => {
        // Given
        const props = {id: 'to'};
        const headerAddress = shallow(<HeaderAddress {...props}/>);

        // When
        const result = headerAddress.find('Autosuggest').props().renderSuggestion('1337');

        // Then
        expect(result).toBe('1337');
      });
      test('onSuggestionSelected, valid suggestion, should add address and clear input value', () => {
        // Given
        const onAddressAdd = jest.fn();
        const props = {id: 'to', onAddressAdd};
        const headerAddress = shallow(<HeaderAddress {...props}/>);
        headerAddress.state().value = '1337';

        // When
        headerAddress.find('Autosuggest').props().onSuggestionSelected({}, {suggestionValue: '1337'});

        // Then
        expect(headerAddress.state().value).toEqual('');
        expect(onAddressAdd).toHaveBeenCalledTimes(1);
      });
    });
    describe('HeaderKeyDown', () => {
      test('HeaderKeyPress, with valid e-mail and Enter key, should trigger addressAdd', () => {
        // Given
        const onAddressAdd = jest.fn();
        const setCustomValidity = jest.fn();
        const props = {id: 'to'};
        const headerAddress = mount(<HeaderAddress {...props} onAddressAdd={onAddressAdd}/>);

        // When
        headerAddress.find('input#to').simulate('keyDown', {
          key: 'Enter', preventDefault: () => { },
          target: {
            id: 'to', value: 'valida@email', validity: {valid: true},
            focus: () => {}, setCustomValidity}
        });

        // Then
        expect(onAddressAdd).toHaveBeenCalledTimes(1);
        expect(setCustomValidity).toHaveBeenCalledTimes(1);
      });
      test('HeaderKeyPress, with valid e-mail and NOT Enter key, should do nothing', () => {
        // Given
        const onAddressAdd = jest.fn();
        const props = {id: 'to'};
        const headerAddress = mount(<HeaderAddress {...props} onAddressAdd={onAddressAdd}/>);

        // When
        headerAddress.find('input#to').simulate('keyDown', {
          key: 'a', preventDefault: () => {}
        });

        // Then
        expect(onAddressAdd).toHaveBeenCalledTimes(0);
      });
      test('HeaderKeyPress, with invalid e-mail and Enter key, should do nothing', () => {
        // Given
        const onAddressAdd = jest.fn();
        const setCustomValidity = jest.fn();
        const reportValidity = jest.fn();
        const props = {id: 'to'};
        const headerAddress = mount(<HeaderAddress {...props} onAddressAdd={onAddressAdd}/>);

        // When
        headerAddress.find('input#to').simulate('keyDown', {
          key: 'Enter',
          target: {id: 'to', value: 'not valid', validity: {valid: false}, setCustomValidity, reportValidity}
        });

        // Then
        expect(onAddressAdd).toHaveBeenCalledTimes(0);
        expect(setCustomValidity).toHaveBeenCalledTimes(2);
        setTimeout(() => expect(reportValidity).toHaveBeenCalledTimes(1));
      });
    });
    describe('HeaderBlur', () => {
      test('HeaderBlur, with valid e-mail, should trigger addressAdd', () => {
        // Given
        const onAddressAdd = jest.fn();
        const props = {id: 'to'};
        const headerAddress = mount(<HeaderAddress {...props} onAddressAdd={onAddressAdd}/>);

        // When
        headerAddress.find('input#to').simulate('blur', {
          target: {id: 'to', value: 'valida@email', validity: {valid: true}}
        });

        // Then
        expect(onAddressAdd).toHaveBeenCalledTimes(1);
      });
      test('HeaderBlur, with invalid e-mail, should do nothing', () => {
        // Given
        const onAddressAdd = jest.fn();
        const setCustomValidity = jest.fn();
        const reportValidity = jest.fn();
        const props = {id: 'to'};
        const headerAddress = mount(<HeaderAddress {...props} onAddressAdd={onAddressAdd}/>);

        // When
        headerAddress.find('input#to').simulate('blur', {
          preventDefault: () => {},
          target: {id: 'to', value: 'not valid', validity: {valid: false}, setCustomValidity, reportValidity}
        });

        // Then
        expect(onAddressAdd).toHaveBeenCalledTimes(0);
        expect(setCustomValidity).toHaveBeenCalledTimes(1);
        setTimeout(() => expect(reportValidity).toHaveBeenCalledTimes(1));
      });
    });
  });
});
