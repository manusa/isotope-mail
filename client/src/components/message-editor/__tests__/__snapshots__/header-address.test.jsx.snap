// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderAddress component test suite Snapshot render, should render HeaderAddress 1`] = `
<div
  className="message-editor__header-address mdc-menu-surface--anchor"
  onClick={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <label>
    BlackLabel12yrs
    :
  </label>
  <div
    className="message-editor__header-chip mdc-chip"
    draggable={true}
    key="0"
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="mdc-chip__text"
    >
      name@email.com
    </div>
    <i
      className="material-icons mdc-chip__icon
               mdc-chip__icon--trailing"
      onClick={[Function]}
    >
      cancel
    </i>
  </div>
  <div
    className="message-editor__header-chip mdc-chip"
    draggable={true}
    key="1"
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="mdc-chip__text"
    >
      other@e-mail.mail
    </div>
    <i
      className="material-icons mdc-chip__icon
               mdc-chip__icon--trailing"
      onClick={[Function]}
    >
      cancel
    </i>
  </div>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "id": "I'mDown",
        "onBlur": [Function],
        "onChange": [Function],
        "onKeyDown": [Function],
        "type": "text",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "container": "message-editor__header-auto-suggest ",
        "suggestion": "mdc-list-item",
        "suggestionHighlighted": "mdc-list-item--activated",
        "suggestionsContainer": "mdc-menu mdc-menu-surface message-editor__header-auto-suggest-menu",
        "suggestionsContainerOpen": "mdc-menu-surface--open",
        "suggestionsList": "mdc-list mdc-list--dense",
      }
    }
  />
</div>
`;
